Add history of code intelligence (#1628)
sourcegraph/about: handbook/engineering/code-intelligence/index.md

- Robust, extensible, and scalable infrastructure to index code across all languages, keep those indexes up-to-date, and efficiently resolve code intelligence queries against all indexed code.

## History

> The origin of this team is the core of Sourcegraph, the first iteration was [srclib](https://github.com/sourcegraph/srclib) a code analysis tool developed by our co-founders [Beyang](https://github.com/beyang) and [Quinn](https://github.com/sqs) in 2015.

Reviewer: The first idea here seems unfinished and unrelated to the second.

Author: What do you mean by unfinished? It's the genesis in short and sweet format - I don't intend for it to be comprehensive.

Reviewer: > The origin of this team is the core of Sourcegraph

It's not clear to me how this should be interpreted. Does this mean src lib was the core of Sourcegraph and also the origin for code intelligence? Or that our team is the core of Sourcegraph?

Author: Oh I see, you mean "our team" as in the people here today were part of original devs?

Reviewer: Yeah, I think it's throwing me at "the first iteration [of what?] ...".

Author: Updated

Reviewer: Thanks! This makes a lot more sense to me.

2016 - We evolved our strategy to utilize LSP (Language Server Protocol) to power the new Sourcegraph code search engine. LSP is a user oriented protocol and requires an implementation of a language server for each programming language.

> 2017 - Over time we started to experience some issues with LSPs; they can be difficult to deploy, slow at runtime, slow community adoption, and slow to develop!

Reviewer 2: ```suggestion
2017 - Over time we started to experience some issues with language servers. They can be difficult to deploy, slow at runtime, slow to adopt by members of their respective language communities, and slow to develop.
```

> 2017 - Over time we started to experience some issues with LSPs; they can be difficult to deploy, slow at runtime, slow community adoption, and slow to develop!

Reviewer: Since this isn't an event or change maybe you should move this text closer to what we did about feeling that pain.

Author: You're right that it isn't an event or change but it was an experience or learning process and I wanted to highlight that.

2018 - We added a new basic code intelligence that is built on search-based heuristics to allow us to provide quick and good enough i.e. "imprecise" support for the [most popular programming languages](https://sourcegraph.com/extensions?query=category%3A%22Programming+languages%22). 

> 2019 - While looking for more efficient alternatives to LSPs we found LSIF (Language Server Index Format) and realized we could bring back good parts of srclib with pre-indexing and pre-computed answers to queries.

Reviewer 2: ```suggestion
2019 - While looking for more efficient alternatives to language servers we found LSIF (Language Server Index Format) which provided an index-based implementation of precise code navigation (similar to srclib) with the advantage of being able to build on top of the work done in the LSP community (LSIF and LSP are sister protocols).
```

> 2019 - While looking for more effecient alternatives to LSPs we found LSIF (Language Server Index Format) and realized we could bring back good parts of srclib with pre-indexing and pre-computed answers to queries.

Reviewer: ```suggestion
2019 - While looking for more efficient alternatives to LSPs we found LSIF (Language Server Index Format) and realized we could bring back good parts of srclib with pre-indexing and pre-computed answers to queries.
```

> 2020 and beyond - Code Intelligence is the platform for all features at Sourcegraph! This team is responsible for how sourcegraph understands code and for providing data to power search, campaigns and code insights. Our API is the point of discovery for the knowledge graph for all source code.

Reviewer 2: ```suggestion
2020 and beyond - The Code Intelligence team is responsible for how Sourcegraph understands code and for providing data that powers semantic search, campaigns and code insights. Our API is the point of discovery for the knowledge graph for all source code.
```
